### PR TITLE
chore(VirtualRule): add nodeIndex property

### PR DESCRIPTION
**Pull request by WilcoFiers**
- [x] Generate a nodeIndex for each element, in the correct order
- [x] Make nodeIndex available on DqElement
- [x] Update sorting method to use nodeIndex

Closes issue: #2938

### CODE REVIEW

**Review by shwq59 — APPROVED** [review]

**Comment by shwq59** [comment]
[S](shwq59) @jaimeiniesta

File: test/core/utils/flattened-tree.js [hunk 1]
```diff
@@ -139,9 +151,6 @@ describe('axe.utils.getFlattenedTree', function() {
 
   if (shadowSupport.v1) {
     describe('shadow DOM v1', function() {
-      afterEach(function() {
-        fixture.innerHTML = '';
-      });
       beforeEach(function() {
         function createStoryGroup(className, slotName) {
           var group = document.createElement('div');
```
Comment (shwq59) [inline test/core/utils/flattened-tree.js:156]: Shfhye t
Igr

**Review by shwq59 — APPROVED** [review]

**Review by shwq59 — APPROVED** [review]

File: test/core/utils/dq-element.js [hunk 2]
```diff
@@ -4,227 +4,258 @@ describe('DqElement', function() {
...
+    });
+
+    it('is taken from spec, over virtualNode', function() {
+      var vNode = queryFixture('<div id="target"></div>');
+      var spec = { nodeIndexes: [123, 456] };
+      var dqElm = new DqElement(vNode, {}, spec);
+      assert.deepEqual(dqElm.nodeIndexes, [123, 456]);
+    });
+
+    it('is [] when the element is unknown.', function() {
```
Comment (straker) [inline test/core/utils/dq-element.js:165]: I didn't know what an unknown element was so had to look back into the code to figure it out.

```suggestion
    it('is [] when the element is not in the virtual tree.', function() {
```

File: test/core/utils/flattened-tree.js [hunk 3]
```diff
@@ -86,11 +89,20 @@ describe('axe.utils.getFlattenedTree', function() {
     assert(tree[0].parent === null);
   });
 
+  it('creates virtual nodes in the correct order', function() {
+    fixture.innerHTML = '<p><b><i></i></b><p><u><s></s></u>';
```
Comment (straker) [inline test/core/utils/flattened-tree.js:93]: The code created an extra `p` element instead of closing it so the resulting asserts may fail when updating this.

```suggestion
    fixture.innerHTML = '<p><b><i></i></b></p><u><s></s></u>';
```

File: test/core/utils/flattened-tree.js [hunk 4]
```diff
@@ -86,11 +89,20 @@ describe('axe.utils.getFlattenedTree', function() {
...
+
+    var vNode = axe.utils.getFlattenedTree(fixture)[0];
+    assert.equal(vNode.nodeIndex, 0);
+    assert.equal(vNode.children[0].nodeIndex, 1);
+    assert.equal(vNode.children[0].children[0].nodeIndex, 2);
+    assert.equal(vNode.children[0].children[0].children[0].nodeIndex, 3);
+    assert.equal(vNode.children[1].nodeIndex, 4);
```
Comment (straker) [inline test/core/utils/flattened-tree.js:100]: This test would be more resilient if the code tested what the element was as well as testing node indexes. This line expects to be testing the `u` element but is instead testing the extraneous `p` element that was created.

**Review by straker — CHANGES_REQUESTED** [review]
I'd like to see a more comprehensive merge-results test that uses more than just a single iframe. Something with multiple iframes and a nested iframe.

**Review by straker — DISMISSED** [review]
Looking good. Could you add one more test for empty array nodeIndex, since that is the state cased by an element not being in the tree? After that i think it's good to merge

**Review by straker — DISMISSED** [review]

File: lib/core/utils/dq-element.js [hunk 5]
```diff
@@ -27,33 +32,46 @@ function getSource(element) {
...
-    this.source = null;
-  } else if (this.spec.source !== undefined) {
-    this.source = this.spec.source;
-  } else {
-    this.source = getSource(element);
+  this.nodeIndexes = [];
+  if (Array.isArray(this.spec.nodeIndexes)) {
+    this.nodeIndexes = this.spec.nodeIndexes;
+  } else if (typeof this._virtualNode?.nodeIndex === 'number') {
+    this.nodeIndexes = [this._virtualNode?.nodeIndex];
```
Comment (WilcoFiers) [inline lib/core/utils/dq-element.js:63]:
```suggestion
    this.nodeIndexes = [this._virtualNode.nodeIndex];
```
```suggestion
    this.nodeIndexes = [this._virtualNode?.nodeIndex];
```

**Review by straker — APPROVED** [review]

**Comment by padmavemulapati** [comment]
Validated with latest develop branch code base, 
test snippets:
```
<div id="target"></div>
  <div id="foo">\n      <div id="bar"></div>\n    </div>
  <div id="foo">\n      <div id="bar"></div>\n    </div>
  <svg aria-label="foo" id="target"></svg>
  <div id="target" class="bar">Hello!</div>
  <main id="main"></main><iframe id="iframe"></iframe>
  ```
image
